### PR TITLE
librbd: Improved support for balanced and localized reads

### PR DESCRIPTION
**Pull request by zhengchengyao**
Fixes: https://tracker.ceph.com/issues/43793
Signed-off-by: Zheng Yin <zhengyin@cmss.chinamobile.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/rbd/rbd-config-ref.rst [hunk 1]
```diff
@@ -15,6 +15,14 @@ Generic IO Settings
 :Default: ``none``
 :Values: ``none``, ``compressible``, ``incompressible``
 
+``rbd read from replica policy``
+
+:Description: read replica policy to send to the OSDs on read operations. If set to `balance`, the OSD will attempt to read the data from primary or replica. If set to `localize`, the OSD will attempt to read local data.
```
Comment (dillaman) [inline doc/rbd/rbd-config-ref.rst]: Nit: 
```suggestion
:Description: policy for determining which OSD will receive read operations. If set to `default`, the primary OSD will always be used for read operations. If set to `balance`, read operations will be sent to a randomly selected OSD within the replica set. If set to `localize`, read operations will be sent to the closest OSD as determined by the CRUSH map. Note: this feature requires the cluster to be configured with a minimum compatible OSD release of Octopus.
```

File: src/librbd/io/CopyupRequest.cc [hunk 2]
```diff
@@ -175,11 +175,12 @@ void CopyupRequest<I>::read_from_parent() {
   if (m_image_ctx->enable_sparse_copyup) {
     ImageRequest<I>::aio_read(
       m_image_ctx->parent, comp, std::move(m_image_extents),
-      ReadResult{&m_copyup_extent_map, &m_copyup_data}, 0, m_trace);
+      ReadResult{&m_copyup_extent_map, &m_copyup_data},
+      m_image_ctx->op_flags, m_trace);
```
Comment (dillaman) [inline src/librbd/io/CopyupRequest.cc]: Nit: unnecessary -- the op flags should only be applied in the `ObectReadRequest` already. Just update `ImageCtx::get_read_flags` appropriately.

File: src/librbd/ImageCtx.cc [hunk 3]
```diff
@@ -795,6 +795,14 @@ class SafeTimerSingleton : public SafeTimer {
       alloc_hint_flags |= librados::ALLOC_HINT_FLAG_INCOMPRESSIBLE;
     }
 
+    op_flags = 0;
+    auto read_policy = config.get_val<std::string>("rbd_read_from_replica_policy");
```
Comment (dillaman) [inline src/librbd/ImageCtx.cc]: You need to verify that the min OSD release Octopus (see `get_min_compatible_osd`)

**Comment by dillaman** [comment]
Can you also create a new teuthology test case where you enable this feature (perhaps in just balance mode for simplicity) during an RBD xfstest/fsx test concurrent w/ the OSD thrasher; and set `ceph osd require-osd-release octopus`?

**Comment by zhengchengyao** [comment]
thanks @dillaman , I add read balance mode test case in test_librbd.

File: src/librbd/io/CopyupRequest.cc [hunk 4]
```diff
@@ -172,14 +172,15 @@ void CopyupRequest<I>::read_from_parent() {
   ldout(cct, 20) << "completion=" << comp << ", "
                  << "extents=" << m_image_extents
                  << dendl;
+  int op_flags = m_image_ctx->get_read_flags(m_image_ctx->parent_md.spec.snap_id);
```
Comment (dillaman) [inline src/librbd/io/CopyupRequest.cc]: Nit: drop this change -- the read flags are not op flags.

File: src/librbd/io/ObjectRequest.cc [hunk 5]
```diff
@@ -186,6 +186,8 @@ ObjectReadRequest<I>::ObjectReadRequest(
   : ObjectRequest<I>(ictx, objectno, offset, len, snap_id, "read",
                      parent_trace, completion),
     m_op_flags(op_flags), m_read_data(read_data), m_extent_map(extent_map) {
+    I *image_ctx = this->m_ictx;
+    m_op_flags = m_op_flags ? m_op_flags : image_ctx->op_flags;
```
Comment (dillaman) [inline src/librbd/io/ObjectRequest.cc]: Nit: drop this change -- the read flags are not op flags.

File: src/librbd/io/ObjectRequest.cc [hunk 6]
```diff
@@ -285,7 +287,8 @@ void ObjectReadRequest<I>::read_parent() {
       this, util::get_image_ctx(image_ctx->parent), AIO_TYPE_READ);
   ImageRequest<I>::aio_read(image_ctx->parent, parent_completion,
                             std::move(parent_extents), ReadResult{m_read_data},
-                            0, this->m_trace);
+                            image_ctx->get_read_flags(image_ctx->parent_md.spec.snap_id),
```
Comment (dillaman) [inline src/librbd/io/ObjectRequest.cc]: Nit: drop this change -- the read flags are not op flags.

File: src/test/librbd/test_librbd.cc [hunk 7]
```diff
@@ -1948,6 +1948,7 @@ TEST_F(TestLibRBD, TestIO)
   uint64_t size = 2 << 20;  
 
   ASSERT_EQ(0, create_image(ioctx, name.c_str(), size, &order));
+  ASSERT_EQ(0, rados_conf_set(_cluster, "rbd_read_from_replica_policy", "balance"));
```
Comment (dillaman) [inline src/test/librbd/test_librbd.cc]: This won't really test anything -- instead you should add a new YAML snippet here [1] that overrides the config setting and ensures the min compatible OSD release is set to Octopus.

[1] https://github.com/ceph/ceph/tree/master/qa/suites/rbd/qemu/features

File: src/librbd/ImageCtx.h [hunk 8]
```diff
@@ -207,6 +207,7 @@ namespace librbd {
     bool clone_copy_on_read;
     bool enable_alloc_hint;
     uint32_t alloc_hint_flags = 0U;
+    uint32_t op_flags = 0U;
```
Comment (dillaman) [inline src/librbd/ImageCtx.h]: Nit: it's an op-flag:
```suggestion
    uint32_t read_flags = 0U;
```

**Comment by zhengchengyao** [comment]
Sorry, I confused op_flags and read_flags, thanks.

File: qa/suites/rbd/qemu/features/readbalance.yaml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    conf:
+      client:
+        rbd read from replica policy: balance
+
+tasks:
+- exec:
+    osd.0:
+      - ceph osd require-osd-release octopus
+      - ceph osd set-require-min-compat-client octopus
+      - for f in `ceph osd pool ls` ; do ceph osd pool set $f pg_autoscale_mode off ; done
```
Comment (dillaman) [inline qa/suites/rbd/qemu/features/readbalance.yaml]: Nit: should be unnecessary

File: qa/suites/rbd/qemu/features/readbalance.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+- ceph.healthy:
```
Comment (dillaman) [inline qa/suites/rbd/qemu/features/readbalance.yaml]: Nit: should be unnecessary

Comment (zhengchengyao) [reply qa/suites/rbd/qemu/features/readbalance.yaml]: I drop it， thanks

**Comment by dillaman** [comment]
jenkins test make check

**Pull request merged**